File: src/main/java/com/fatec/destino/repository/pacote/PacoteRepository.java
package com.fatec.destino.repository.pacote;

import com.fatec.destino.dto.pacote.PacoteResponseDTO;
import com.fatec.destino.model.pacote.Pacote;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

@Repository
public interface PacoteRepository extends JpaRepository<Pacote, Long> {

        @Query("SELECT p from Pacote p JOIN p.ofertas o where o.status = com.fatec.destino.util.model.pacote.OfertaStatus.EMANDAMENTO")
        Page<Pacote> encontrePacotes(Pageable pageable);

        @Query("SELECT p from Pacote p where p.nome = :nome")
        List<Pacote> procurePeloNome(@Param("nome") String nome);

        // Busca os pacotes mais vendidos com base no número de compras
        @Query("SELECT c.oferta.pacote FROM Compra c GROUP BY c.oferta.pacote ORDER BY COUNT(c) DESC")
        List<Pacote> procuraPacotesMaisVendidos();

        @Query("SELECT p FROM Pacote p JOIN p.ofertas o WHERE o.status = 'EMANDAMENTO' OR o.status = 'CONCLUIDO' ORDER BY o.status DESC, o.inicio DESC")
        List<Pacote> findPacotesPublicos();

        @Query("SELECT p FROM Pacote p JOIN p.ofertas o WHERE o.status = 'EMANDAMENTO' " +
                        "AND (:nome IS NULL OR LOWER(p.nome) LIKE LOWER(CONCAT('%', :nome, '%'))) " +
                        "AND (:precoMax IS NULL OR o.preco <= :precoMax)")
        Page<Pacote> buscarComFiltros(
                        @Param("nome") String nome,
                        @Param("precoMax") BigDecimal precoMax,
                        Pageable pageable);

        // Busca exata para a página de detalhes (URL amigável)
        Optional<Pacote> findByNome(String nome);

        // --- Métodos de Verificação de Integridade ---
        boolean existsByHotelId(long id);

        boolean existsByTransporteId(long id);

        // --- Dashboard ---
        @Query("SELECT o.status, COUNT(p) FROM Pacote p JOIN p.ofertas o GROUP BY o.status")
        List<Object[]> countPacotesByStatus();

        @Query("SELECT t.meio, COUNT(p) FROM Pacote p JOIN p.transporte t GROUP BY t.meio")
        List<Object[]> countPacotesByTransporteMeio();

        @Query("SELECT MONTH(o.inicio), COUNT(p) FROM Pacote p JOIN p.ofertas o WHERE o.status = 'CONCLUIDO' AND YEAR(o.inicio) = :ano GROUP BY MONTH(o.inicio) ORDER BY MONTH(o.inicio) ASC")
        List<Object[]> countViagensConcluidasByMonth(@Param("ano") int ano);
}

File: src/main/kotlin/com/fatec/destino/repository/pacote/PacoteRepository.kt
package com.fatec.destino.repository.pacote

import com.fatec.destino.dto.pacote.PacoteResponseDTO
import com.fatec.destino.model.pacote.Pacote
import com.fatec.destino.util.repository.MeioCount
import com.fatec.destino.util.repository.MonthCount
import com.fatec.destino.util.repository.StatusCount
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository
import java.math.BigDecimal

@Repository
interface PacoteRepository : JpaRepository<Pacote, Long> {

    @Query("""
        SELECT new com.fatec.destino.dto.pacote.PacoteResponseDTO(
        p.id, 
        p.nome, 
        CAST(p.preco AS int), 
        p.hotel.nome, 
        p.transporte.empresa
    ) 
        FROM Pacote p WHERE p.nome = :nome
    """)
    fun procurePeloNome(nome: String): List<PacoteResponseDTO>

    @Query("""
    SELECT new com.fatec.destino.dto.pacote.PacoteResponseDTO(
        c.viagem.pacote.id, 
        c.viagem.pacote.nome, 
        CAST(c.viagem.pacote.preco AS int), 
        c.viagem.pacote.hotel.nome, 
        c.viagem.pacote.transporte.empresa
    ) 
    FROM Compra c 
    GROUP BY c.viagem.pacote.id, c.viagem.pacote.nome, c.viagem.pacote.preco, c.viagem.pacote.hotel.nome, c.viagem.pacote.transporte.empresa 
    ORDER BY COUNT(c.viagem.pacote) DESC
""")
    fun procuraPacotesMaisVendidos(): List<PacoteResponseDTO>

    fun findByNome(nome: String): Pacote?

    fun existsByHotelId(id: Long): Boolean

    fun existsByTransporteId(id: Long): Boolean
}